### PR TITLE
[PyTorch] Port ExecuTorch bfdot improvement back to ATen BlasKernel, Try #2

### PR DESCRIPTION
**Pull request by swolchok**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #137723
* #137661
* #137426
* #137722
* __->__ #137377

ExecuTorch's fork of BlasKernel.cpp grew bfdot support, complete with demonstration that it helps. Port it back to PyTorch. First attempt was https://github.com/pytorch/pytorch/pull/136331 .

Differential Revision: [D63923166](https://our.internmc.facebook.com/intern/diff/D63923166/)

### CODE REVIEW

**Comment by albanD** [comment]
What's the difference with the previous attempt?

File: aten/src/ATen/native/BlasKernel.cpp [hunk 1]
```diff
@@ -455,67 +456,183 @@ static C10_ALWAYS_INLINE float32x4_t f32_fma_bf16(float32x4_t a, uint16x4_t b, u
...
+    const auto* vec2_ = vec2 + j;
+    c10::ForcedUnroll<kF32RegisterPairsPerIteration>{}([vec1_, vec2_, &sum](auto k) C10_ALWAYS_INLINE_ATTRIBUTE {
+      dot_with_fp32_arith_main_inner_loop_no_bfdot(vec1_, vec2_, sum, k);
+    });
   }
-  return reducedSum;
+  return reduce(sum);
 }
 
+// NOTE: The first attempt at landing BFDOT support with
```
Comment (swolchok) [inline aten/src/ATen/native/BlasKernel.cpp:581]: @albanD here is the difference

**Comment by swolchok** [comment]
will fix https://hud.pytorch.org/pr/pytorch/pytorch/137377#31249893287

File: aten/src/ATen/native/BlasKernel.cpp (hunk 1 above)
Comment (albanD) [reply aten/src/ATen/native/BlasKernel.cpp:581]: Ho thanks for the pointer!
Should we undef DOT_WITH_FP32_ARITH_TAIL_AFTER_MAIN_LOOP_BODY ?

Comment (swolchok) [reply aten/src/ATen/native/BlasKernel.cpp]: this is a .cpp file, so the macro isn't going to leak anywhere and it's not particularly necessary, but sure I can do that.

Comment (swolchok) [reply aten/src/ATen/native/BlasKernel.cpp]: done

**Comment by swolchok** [comment]
CI is 100% green; please review

**Review by malfet — APPROVED** [review]

**Comment by malfet** [comment]
If CI is green, than sure, LGTM

**Pull request closed without merging**